dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchMemoryRecordService.cs

```diff
@@ -0,0 +1,357 @@
+    private static readonly HashSet<Type> s_supportedKeyTypes = new()
+    {
+        typeof(string)
+    };
+
+    /// <summary>A set of types that vectors on the provided model may have.</summary>
+    private static readonly HashSet<Type> s_supportedVectorTypes = new()
+    {
+        typeof(ReadOnlyMemory<float>),
+        typeof(ReadOnlyMemory<float>?)
+    };
+
+    /// <summary>Azure AI Search client that can be used to manage the list of indices in an Azure AI Search Service.</summary>
+    private readonly SearchIndexClient _searchIndexClient;
+
+    /// <summary>The name of the collection to use with this store if none is provided for any individual operation.</summary>
+    private readonly string _defaultCollectionName;
+
+    /// <summary>The name of the key field for the collections that this class is used with.</summary>
+    private readonly string _keyPropertyName;
+
+    /// <summary>Azure AI Search clients that can be used to manage data in an Azure AI Search Service index.</summary>
+    private readonly ConcurrentDictionary<string, SearchClient> _searchClientsByIndex = new();
+
+    /// <summary>Optional configuration options for this class.</summary>
+    private readonly AzureAISearchMemoryRecordServiceOptions<TDataModel> _options;
+
+    /// <summary>The names of all non vector fields on the current model.</summary>
+    private readonly List<string> _nonVectorPropertyNames;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AzureAISearchMemoryRecordService{TDataModel}"/> class.
+    /// </summary>
+    /// <param name="searchIndexClient">Azure AI Search client that can be used to manage the list of indices in an Azure AI Search Service.</param>
+    /// <param name="defaultCollectionName">The name of the collection to use with this store if none is provided for any individual operation.</param>
+    /// <param name="options">Optional configuration options for this class.</param>
+    /// <exception cref="ArgumentNullException">Thrown when <paramref name="searchIndexClient"/> is null.</exception>
+    /// <exception cref="ArgumentException">Thrown when <paramref name="defaultCollectionName"/> is null or whitespace.</exception>
+    public AzureAISearchMemoryRecordService(SearchIndexClient searchIndexClient, string defaultCollectionName, AzureAISearchMemoryRecordServiceOptions<TDataModel>? options = default)
+    {
+        // Verify.
+        Verify.NotNull(searchIndexClient);
+        Verify.NotNullOrWhiteSpace(defaultCollectionName);
+
+        // Assign.
+        this._searchIndexClient = searchIndexClient;
+        this._defaultCollectionName = defaultCollectionName;
+        this._options = options ?? new AzureAISearchMemoryRecordServiceOptions<TDataModel>();
+
+        // Verify custom mapper.
+        if (this._options.MapperType == AzureAISearchMemoryRecordMapperType.JsonObjectCustomerMapper && this._options.JsonObjectCustomMapper is null)
```

Comment:
I was thinking that there may be more built in types, for example where we want to support first party models that require mapping other than what the default mapper can do.  The models we have today and any kind of generic model that just uses dictionaries would definitely require this type of custom mapper.  Of course, they could always be passed in via the constructor as just another custom mapper, instead of being added to the enum, but since not all of the mappers are separable like that, it seems wrong to have some of our 1st party mappers injected and some being just a config option.